### PR TITLE
BUG: inverting object dtype containing bools

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Motivation: when trying to deprecate the inplace keyword in `where` in favor of using `mask` with a flipped condition, i found some tests failed because we pass a object-dtype frame of bools and inverting that frame gives weird results.  This is because the python behavior is weird:

```
>>> ~True
-2
>>> ~False
-1
```

This has the impact that if we have a bool-dtype Series and invert it, the behavior doesn't pointwise-match the object-dtype behavior.  This PR changes the object-dtype behavior to what I think is more reasonable. 

If someone were to complain about this diverging from the python/numpy behavior, I'd understand.

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -598,6 +598,43 @@ def maybe_booleans_to_slice(ndarray[uint8_t, ndim=1] mask):
...
+        Py_ssize_t i, N = values.size
+        object val, res_val
+        ndarray result = cnp.PyArray_EMPTY(values.ndim, values.shape, cnp.NPY_OBJECT, 0)
+        object[::1] res_flat = result.ravel()  # should NOT be a copy
+        cnp.flatiter it = cnp.PyArray_IterNew(values)
+
+    for i in range(N):
+
+        # Analogous to: val = values[i]
+        val = PyArray_GETITEM(values, PyArray_ITER_DATA(it))
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:613]: Does this make any difference compared to the normal array access? I would be surprised if so - the annotate functionality is a good way to check

Comment (jbrockmendel) [reply pandas/_libs/lib.pyx:613]: this is the pattern we use to handle arbitrary-ndim ndarrays

**Comment by mroeschke** [comment]
Was there an existing issue about this? I found https://github.com/pandas-dev/pandas/issues/51567 but not sure if this case was also specifically discussed somewhere

**Comment by jbrockmendel** [comment]
> Was there an existing issue about this?

I dont think we have one, but there is https://github.com/numpy/numpy/issues/23926, also i think i mentioned it on slack at some point

File: doc/source/whatsnew/v2.1.0.rst [hunk 2]
```diff
@@ -698,7 +698,8 @@ Numeric
 - Bug in :meth:`Series.corr` and :meth:`Series.cov` raising ``AttributeError`` for masked dtypes (:issue:`51422`)
 - Bug in :meth:`Series.median` and :meth:`DataFrame.median` with object-dtype values containing strings that can be converted to numbers (e.g. "2") returning incorrect numeric results; these now raise ``TypeError`` (:issue:`34671`)
 - Bug in :meth:`Series.sum` converting dtype ``uint64`` to ``int64`` (:issue:`53401`)
-
+- Bug in operator in inverting an object-dtype :class:`DataFrame` column or :class:`Series` containing ``bool`` objects giving integer results instead of flipped bools (:issue:`54569`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:701]: Do you think this is a large enough bug to go in 2.1?

**Comment by lithomas1** [comment]
Can we raise/do nothing instead to follow future Python versions? Python 3.12 deprecated this.

https://docs.python.org/3.12/whatsnew/3.12.html#:~:text=The%20bitwise%20inversion,gh%2D103487.)

**Comment by jbrockmendel** [comment]
I wasnt aware of that python deprecation.  That makes doing nothing more attractive.  But then that'll mean for bool dtype the vector behavior wont match the pointwise (raising) behavior.  Deprecating `~` for bool dtype would seem really weird though.  do we just live with the difference?

(note to self: open an issue about deprecating Timestamp utcfromtimestamp, utcnow as they are deprecated in the stdlib)

**Pull request closed without merging**